qiskit_nature/properties/second_quantization/electronic/dipole_moment.py

```diff
@@ -86,40 +119,93 @@ def from_driver_result(cls, result: DriverResult) -> Optional["TotalDipoleMoment
+            if mo_ints[0] is not None:
+                integrals.append(OneBodyElectronicIntegrals(ElectronicBasis.MO, mo_ints))
+
+            return DipoleMoment(axis, integrals, shift=energy_shift)
 
         return cls(
-            {
-                "x": dipole_along_axis(
-                    "x", (qmol.x_dip_mo_ints, qmol.x_dip_mo_ints_b), qmol.x_dip_energy_shift
+            [
+                dipole_along_axis(
+                    "x",
+                    (qmol.x_dip_ints, None),
+                    (qmol.x_dip_mo_ints, qmol.x_dip_mo_ints_b),
+                    qmol.x_dip_energy_shift,
                 ),
-                "y": dipole_along_axis(
-                    "y", (qmol.y_dip_mo_ints, qmol.y_dip_mo_ints_b), qmol.y_dip_energy_shift
+                dipole_along_axis(
+                    "y",
+                    (qmol.y_dip_ints, None),
+                    (qmol.y_dip_mo_ints, qmol.y_dip_mo_ints_b),
+                    qmol.y_dip_energy_shift,
                 ),
-                "z": dipole_along_axis(
-                    "z", (qmol.z_dip_mo_ints, qmol.z_dip_mo_ints_b), qmol.z_dip_energy_shift
+                dipole_along_axis(
+                    "z",
+                    (qmol.z_dip_ints, None),
+                    (qmol.z_dip_mo_ints, qmol.z_dip_mo_ints_b),
+                    qmol.z_dip_energy_shift,
                 ),
-            },
+            ],
             dipole_shift={
                 "nuclear dipole moment": cast(
                     DipoleTuple, tuple(d_m for d_m in qmol.nuclear_dipole_moment)
                 ),
             },
+            reverse_dipole_sign=qmol.reverse_dipole_sign,
         )
 
     def second_q_ops(self) -> List[FermionicOp]:
         """Returns a list of dipole moment operators along all Cartesian axes."""
-        return [dip.second_q_ops()[0] for dip in self._dipole_axes.values()]
+        return [dip.second_q_ops()[0] for dip in self._properties.values()]
+
+    def interpret(self, result: EigenstateResult) -> None:
+        """Interprets an :class:~qiskit_nature.result.EigenstateResult in this property's context.
+
+        Args:
+            result: the result to add meaning to.
+        """
+        result.nuclear_dipole_moment = self._dipole_shift.pop("nuclear dipole moment", None)
+        result.reverse_dipole_sign = self._reverse_dipole_sign
+        result.computed_dipole_moment = []
+        result.extracted_transformer_dipoles = []
+
+        if not isinstance(result.aux_operator_eigenvalues, list):
+            aux_operator_eigenvalues = [result.aux_operator_eigenvalues]
+        else:
+            aux_operator_eigenvalues = result.aux_operator_eigenvalues  # type: ignore
+        for aux_op_eigenvalues in aux_operator_eigenvalues:
+            if aux_op_eigenvalues is None:
+                continue
+
+            if len(aux_op_eigenvalues) >= 6:
+                dipole_moment = []
+                for moment in aux_op_eigenvalues[3:6]:
```

Comment:
We really should go over to some dictionary i/o where indexing is not required. When it was more or less in one file that created the operators and interpreted them it was kinda ok.

Reply:
Discussion continues [here](https://github.com/Qiskit/qiskit-nature/pull/263#discussion_r664583546)